BaseLib/Functional.h

```diff
@@ -0,0 +1,214 @@
+                     IndexedPlacedPlaceholder<Indices>::value...);
+}
+
+template <int... Indices, typename Object, typename MethodClass,
+          typename ReturnType, typename... Args>
+std::function<ReturnType(Args...)> easyBind_inner(
+    ReturnType (MethodClass::*method)(Args...), Object&& obj,
+    IntegerSequence<Indices...>)
+{
+    return easyBind_innermost<Indices...>(method, std::forward<Object>(obj));
+}
+
+template <int... Indices, typename Object, typename MethodClass,
+          typename ReturnType, typename... Args>
+std::function<ReturnType(Args...)> easyBind_inner(
+    ReturnType (MethodClass::*method)(Args...) const, Object&& obj,
+    IntegerSequence<Indices...>)
+{
+    return easyBind_innermost<Indices...>(method, std::forward<Object>(obj));
+}
+
+/*! Deduces the signature of the call operator of class \c T.
+ *
+ * The matching type of std::function is provided as the member type
+ * \c FunctionType.
+ *
+ * \see http://stackoverflow.com/a/7943765
+ */
+template <typename T>
+struct FunctionTraits
+    : public FunctionTraits<decltype(&std::decay<T>::type::operator())> {
+};
+
+template <typename Object, typename ReturnType, typename... Args>
+struct FunctionTraits<ReturnType (Object::*)(Args...)> {
+    using FunctionType = std::function<ReturnType(Args...)>;
+};
+
+template <typename Object, typename ReturnType, typename... Args>
+struct FunctionTraits<ReturnType (Object::*)(Args...) const> {
+    using FunctionType = std::function<ReturnType(Args...)>;
+};
+
+}  // namespace detail
+
+/*! Convenience wrapper for std::bind().
+ *
+ * This function binds the member function pointer \c member of class \c Object
+ * to the instance \c obj of this class and wraps the result in a std::function
+ * with matching signature.
+ *
+ * The result of this function can be used, e.g., to deduce the signature of the
+ * \c method (which is not possible with the result of std::bind).
+ *
+ * Example:
+ * \code{.cpp}
+ * using std::placeholders;
+ * Object some_object;
+ *
+ * auto f_bind = std::function<ReturnType(Arg1, Arg2, Arg3>(
+ *                  std::bind(&Object::methodWithThreeArguments,
+ *                      std::ref(some_object), _1, _2, _3);
+ *
+ * auto f_easy = easyBind(&Object::methodWithThreeArguments, some_object);
+ * \endcode
+ *
+ * In the example the expressions creating \c f_bind and \c f_easy are
+ * equivalent.
+ *
+ * \note
+ * There is one difference between the behaviour of std::bind and the one of
+ * easyBind: In easyBind \c obj is never copied, instead it will be referenced.
+ * This is in contrast to the behaviour of std::bind, and has been chosen in
+ * order to prevent accidental copies.
+ */
+template <typename Object, typename MethodClass, typename ReturnType,
+          typename... Args>
+typename std::enable_if<
+    std::is_same<MethodClass,
+                 typename std::remove_cv<typename std::remove_pointer<
+                     typename std::decay<Object>::type>::type>::type>::value,
```

Comment:
Thanks for clarification.

Just for general education: With clang-3.8.0 removing the first (around line 178) `remove_cv` is working fine, but removing the second one (around line 193) also causes a compilation error like:
```
[ 79%] Building CXX object Tests/CMakeFiles/testrunner.dir/BaseLib/TestFunctional.cpp.o
ogs/Tests/BaseLib/TestFunctional.cpp:148:24: error: no matching function for call to 'easyBind'
        auto fpc_get = BaseLib::easyBind(&A::getValue, apc);
                       ^~~~~~~~~~~~~~~~~
ogs/BaseLib/Functional.h:181:1: note: candidate template ignored: could not match 'type-parameter-0-2 (type-parameter-0-3...)' against 'double () const'
easyBind(ReturnType (MethodClass::*method)(Args...), Object&& obj)
^
ogs/BaseLib/Functional.h:192:5: note: candidate template ignored: disabled by 'enable_if' [with Object = const A *&, MethodClass = A, ReturnType = double, Args = <>]
    std::is_same<MethodClass, typename std::remove_pointer<typename std::decay<
    ^
ogs/BaseLib/Functional.h:207:55: note: candidate function template not viable: requires single argument 'obj', but 2 arguments were provided
typename detail::FunctionTraits<Object>::FunctionType easyBind(Object&& obj)
```

So maybe the tests are not testing all the combinations...